package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,331 @@
+
+
+References
+----------
+
+.. [Maginn2019] Maginn, E. J., Messerly, R. A., Carlson, D. J.; Roe, D. R., Elliott, J. R. Best Practices for Computing Transport Properties 1. Self-Diffusivity and Viscosity from Equilibrium Molecular Dynamics [Article v1.0]. Living J. Comput. Mol. Sci. 2019, 1 (1).
+
+
+
+Classes and Functions
+---------------------
+
+.. autoclass:: EinsteinMSD
+    :members:
+
+"""
+
+from __future__ import division, absolute_import
+
+import numpy as np
+import logging
+import tidynamics
+from ..due import due, Doi
+from .base import AnalysisBase
+
+due.cite(Doi("10.21105/joss.00877"),
+         description="Mean Squared Displacements with tidynamics",
+         path="MDAnalysis.analysis.msd",
+         cite_module=True)
+due.cite(Doi("10.1051/sfn/201112010"),
+         description="FCA fast correlation algorithm",
+         path="MDAnalysis.analysis.msd",
+         cite_module=True)
+del Doi
+
+class EinsteinMSD(AnalysisBase):
+    r"""Class to calculate Mean Squared Displacement by the Einstein relation.
+    
+    Attributes
+    ----------
+    dim_fac : int
+        Dimensionality :math:`d` of the MSD.
+    timeseries : :class:`np.ndarray`
+        The averaged MSD with respect to lag-time.
+    msd_per_particle : :class:`np.ndarray`
+        The MSD of each individual particle with respect to lag-time.
+    n_frames : int
+        Number of frames in trajectory.
+    n_particles : int
+        Number of particles MSD was calculated over.
+
+    """
+
+    def __init__(self, u, select=None, msd_type='xyz', fft=True, **kwargs):
+        r"""
+        Parameters
+        ----------
+        u : Universe
+            An MDAnalysis :class:`Universe`.
+        selection : str
+            An MDAnalysis selection string. Defaults to `None`.
+        msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Desired dimensions to be included in the MSD. Defaults to 'xyz'.
+        fft : bool
+            Use a fast FFT based algorithm for computation of the MSD.
+            Otherwise, use the simple "windowed" algorithm. Defaults to ``True``.
+
+        """
+        self.u = u
+        
+        super(EinsteinMSD, self).__init__(self.u.trajectory, **kwargs)
+
+        #args
+        self.select = select
+        self.msd_type = msd_type
+        self.fft = fft
+        
+        #local
+        self._dim = None
+        self._position_array = None
+        self._atoms = None
+        
+        #indexing
+        self._n_frames = 0
+        self._n_particles = 0
+
+        #result
+        self.dim_fac = 0
+        self.timeseries = None
+        self.msds_by_particle = None
+
+    def _prepare(self):
+        self._parse_msd_type()
+        self._atoms = self.u.select_atoms(self.select)
+        self._n_frames = len(self.u.trajectory)
```

Comment:
This is not quite correct: `run(step=10)` will only keep 1/10 of the frames.

There should be a way to get this information somehow. Ultimately, [`check_slice_indices()`](https://www.mdanalysis.org/mdanalysis/documentation_pages/coordinates/base.html#MDAnalysis.coordinates.base.ReaderBase.check_slice_indices) is run somewhere.

Reply:
`start` and `stop` will also change the size

Reply:
When `_prepare()` is run, you already have `self.start`, `self.stop`, and `self.step` available, see https://github.com/MDAnalysis/mdanalysis/blob/6eb8f6812883db30f200ae9b2f73fe96f2227a3b/package/MDAnalysis/analysis/base.py#L186-L188 where `_setup_frames()` is run just before `_prepare()`.
```suggestion
        self._n_frames = len(self.u.trajectory[self.start, self.stop, self.step])
```
(Will be fast because the trajectory iterator is smart like `range()`.)

Reply:
`n_frames` is already an attribute that is also set in `_setup_frames()`.

https://github.com/MDAnalysis/mdanalysis/blob/6eb8f6812883db30f200ae9b2f73fe96f2227a3b/package/MDAnalysis/analysis/base.py#L145

Reply:
Thanks for pointing it out... when I saw it, too, you beat me to it.